### ISSUE TITLE
test_logging leaks a "dangling" thread

### ISSUE THREAD

**Issue by vstinner** [python/cpython#74317]
BPO | [30131](https://bugs.python.org/issue30131)
--- | :---
Nosy | @vstinner, @zhangyangyu
PRs | <li>python/cpython#1275</li><li>python/cpython#1298</li><li>python/cpython#1317</li><li>python/cpython#1318</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-05-02.13:55:46.854>
created_at = <Date 2017-04-21.16:12:01.929>
labels = ['3.7', 'tests', 'performance']
title = 'test_logging leaks a "dangling" thread'
updated_at = <Date 2017-05-02.13:55:46.853>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-05-02.13:55:46.853>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-05-02.13:55:46.854>
closer = 'vstinner'
components = ['Tests']
creation = <Date 2017-04-21.16:12:01.929>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 30131
keywords = []
message_count = 9.0
messages = ['292058', '292241', '292343', '292349', '292395', '292424', '292504', '292505', '292757']
nosy_count = 2.0
nosy_names = ['vstinner', 'xiang.zhang']
pr_nums = ['1275', '1298', '1317', '1318']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'resource usage'
url = 'https://bugs.python.org/issue30131'
versions = ['Python 3.5', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
Example on Windows from AppVeyor:

https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.1402

Warning -- threading.\_dangling was modified by test_logging
  Before: \<_weakrefset.WeakSet object at 0x027CBE30\>
  After:  \<_weakrefset.WeakSet object at 0x027CBFF0\>

I also saw this warning on FreeBSD buildbots.

**Comment by vstinner** [comment]
New changeset 4bcfa3a2363c30d406c6826e291b30c33a649d18 by Victor Stinner in branch 'master':
bpo-30131: Cleanup threads in test_logging (bpo-1275)
https://github.com/python/cpython/commit/4bcfa3a2363c30d406c6826e291b30c33a649d18

**Comment by vstinner** [comment]
After stressing test_logging during 2 days on my FreeBSD VM, it seems like I found the root bug: see https://github.com/python/cpython/pull/1298

**Comment by vstinner** [comment]
New changeset 8ca2f2faefa8dba323a2e4c4b86efb633d7a53cf by Victor Stinner in branch 'master':
bpo-30131: test_logging now joins queue threads (bpo-1298)
https://github.com/python/cpython/commit/8ca2f2faefa8dba323a2e4c4b86efb633d7a53cf

**Comment by zhangyangyu** [comment]
Does this need to be backported?

**Comment by vstinner** [comment]
> Does this need to be backported?

I checked: Python 3.5 and 3.6 have the bug, but Python 2.7 is not affected (it doesn't have the unit test which leaks resources). I created PRs.

**Comment by vstinner** [comment]
New changeset cb21f5f3d2005b891df0b76e3fc4d4676e285f21 by Victor Stinner in branch '3.6':
bpo-30131: test_logging now joins queue threads (bpo-1298) (bpo-1317)
https://github.com/python/cpython/commit/cb21f5f3d2005b891df0b76e3fc4d4676e285f21

**Comment by vstinner** [comment]
New changeset d855e53eabca0e8b2b39d25670c37bbdb4ab2e4c by Victor Stinner in branch '3.5':
bpo-30131: test_logging now joins queue threads (bpo-1298) (bpo-1318)
https://github.com/python/cpython/commit/d855e53eabca0e8b2b39d25670c37bbdb4ab2e4c

**Comment by vstinner** [comment]
Ok, this bug is now fixed in 3.5, 3.6 and master (3.7).
